### ISSUE TITLE
An ElectrodeTableRegion with non-default 'name' causes error on read

### ISSUE THREAD

**Issue by tjd2002** [NeurodataWithoutBorders/pynwb#679]
## Bug

By default an ElectrodeTableRegion (ETR) is created with a name of 'electrodes'. If any other name is specified in the`nwbfile.create_electrode_table_region` method, then an ElectricalTimeSeries referencing this ETR can be written to disk without error or warning, but causes an error on read. (`TypeError: missing argument 'electrodes'`; See repro code and error message below.)

**It would seem important to understand why this condition is only detected on read--ideally the same checks should have caught it at write time as well?**

### Steps to Reproduce
```python
#!/usr/bin/env python3
##########################################
## Boilerplate NWBFile creation...
import pynwb
from datetime import datetime
from dateutil.tz import tzlocal
nwb_filename = '/Users/tjd/tmp/ETR_test.nwb'
source = 'hypothetical source'
nwbf = pynwb.NWBFile(source=source,
               session_description='test electrode_table_region',
               identifier='ETR_test',
               session_start_time=datetime.now(tzlocal()))
electrode_group = nwbf.create_electrode_group(name='electrode0',
                                              source=source,
                                              description='example',
                                              location='somewhere',
                                              device=nwbf.create_device(name='dev0', source=source))
nwbf.add_electrode(x=1.0, y=2.0, z=3.0, imp=0.,
                   location='brain', filtering='none',
                   group=electrode_group)
# ... end boilerplate
##########################################


##########################################
##########################################
## These two are equivalent and both work fine
ETR_default_name     = nwbf.create_electrode_table_region([0], 'First electrode')
ETR_named_electrodes = nwbf.create_electrode_table_region([0], 'First electrode', 
                                                          name='electrodes')
##########################################
## This one triggers bug if used below
ETR_other_name      = nwbf.create_electrode_table_region([0], 'First electrode',
                                                          name='my_name')
##########################################
##########################################

ephys_ts = pynwb.ecephys.ElectricalSeries(
                        name='ETR_test',
                        source=source,
                        data=[0],
                        electrodes=ETR_other_name, #### <- ETR USED HERE
                        timestamps=[1])

nwbf.add_acquisition(ephys_ts)
with pynwb.NWBHDF5IO(nwb_filename, mode='w') as iow:
    iow.write(nwbf)

##########################################
## ERROR ON READ
io = pynwb.NWBHDF5IO(nwb_filename, mode='r')
nwbf_read = io.read()
##########################################
```
Results in the error below:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/Src/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    957         try:
--> 958             obj = cls(**kwargs)
    959             obj.container_source = builder.source

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    372                         msg = ', '.join(parse_err)
--> 373                         raise_from(ExceptionType(msg), None)
    374 

~/conda/envs/pynwb/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

TypeError: missing argument 'electrodes'
```

<details> <summary>[Click to expand rest of error message]</summary><p>

```python
The above exception was the direct cause of the following exception:

Exception                                 Traceback (most recent call last)
<ipython-input-9-c7346c02b572> in <module>
     46 ## ERROR ON READ
     47 io = pynwb.NWBHDF5IO(nwb_filename, mode='r')
---> 48 nwbf_read = io.read()

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    373                         raise_from(ExceptionType(msg), None)
    374 
--> 375                 return func(self, **parsed['args'])
    376         else:
    377             def func_call(*args, **kwargs):

~/Src/pynwb/src/pynwb/form/backends/io.py in read(self, **kwargs)
     31     def read(self, **kwargs):
     32         f_builder = self.read_builder()
---> 33         container = self.__manager.construct(f_builder)
     34         return container
     35 

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    373                         raise_from(ExceptionType(msg), None)
    374 
--> 375                 return func(self, **parsed['args'])
    376         else:
    377             def func_call(*args, **kwargs):

~/Src/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    187         result = self.__containers.get(builder_id)
    188         if result is None:
--> 189             result = self.__type_map.construct(builder, self)
    190             parent_builder = self.__get_parent_dt_builder(builder)
    191             if parent_builder is not None:

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    373                         raise_from(ExceptionType(msg), None)
    374 
--> 375                 return func(self, **parsed['args'])
    376         else:
    377             def func_call(*args, **kwargs):

~/Src/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
   1377                              % str(container.__class__.__name__))  # noqa: F821
   1378         else:
-> 1379             return attr_map.construct(builder, build_manager)
   1380 
   1381     @docval({"name": "container", "type": Container, "doc": "the container to convert to a Builder"},

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    373                         raise_from(ExceptionType(msg), None)
    374 
--> 375                 return func(self, **parsed['args'])
    376         else:
    377             def func_call(*args, **kwargs):

~/Src/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    932         cls = manager.get_cls(builder)
    933         # gather all subspecs
--> 934         subspecs = self.__get_subspec_values(builder, self.spec, manager)
    935         # get the constructor argument that each specification corresponds to
    936         const_args = dict()

~/Src/pynwb/src/pynwb/form/build/map.py in __get_subspec_values(self, builder, spec, manager)
    874                         ret[subspec] = self.__flatten(sub_builder, subspec, manager)
    875             # now process groups and datasets
--> 876             self.__get_sub_builders(groups, spec.groups, manager, ret)
    877             self.__get_sub_builders(datasets, spec.datasets, manager, ret)
    878         elif isinstance(spec, DatasetSpec):

~/Src/pynwb/src/pynwb/form/build/map.py in __get_sub_builders(self, sub_builders, subspecs, manager, ret)
    914                 if dt is None:
    915                     # recurse
--> 916                     ret.update(self.__get_subspec_values(sub_builder, subspec, manager))
    917                 else:
    918                     ret[subspec] = manager.construct(sub_builder)

~/Src/pynwb/src/pynwb/form/build/map.py in __get_subspec_values(self, builder, spec, manager)
    874                         ret[subspec] = self.__flatten(sub_builder, subspec, manager)
    875             # now process groups and datasets
--> 876             self.__get_sub_builders(groups, spec.groups, manager, ret)
    877             self.__get_sub_builders(datasets, spec.datasets, manager, ret)
    878         elif isinstance(spec, DatasetSpec):

~/Src/pynwb/src/pynwb/form/build/map.py in __get_sub_builders(self, sub_builders, subspecs, manager, ret)
    906                 sub_builder = builder_dt.get(dt)
    907                 if sub_builder is not None:
--> 908                     sub_builder = self.__flatten(sub_builder, subspec, manager)
    909                     ret[subspec] = sub_builder
    910             else:

~/Src/pynwb/src/pynwb/form/build/map.py in __flatten(self, sub_builder, subspec, manager)
    919 
    920     def __flatten(self, sub_builder, subspec, manager):
--> 921         tmp = [manager.construct(b) for b in sub_builder]
    922         if len(tmp) == 1 and not subspec.is_many():
    923             tmp = tmp[0]

~/Src/pynwb/src/pynwb/form/build/map.py in <listcomp>(.0)
    919 
    920     def __flatten(self, sub_builder, subspec, manager):
--> 921         tmp = [manager.construct(b) for b in sub_builder]
    922         if len(tmp) == 1 and not subspec.is_many():
    923             tmp = tmp[0]

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    373                         raise_from(ExceptionType(msg), None)
    374 
--> 375                 return func(self, **parsed['args'])
    376         else:
    377             def func_call(*args, **kwargs):

~/Src/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    187         result = self.__containers.get(builder_id)
    188         if result is None:
--> 189             result = self.__type_map.construct(builder, self)
    190             parent_builder = self.__get_parent_dt_builder(builder)
    191             if parent_builder is not None:

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    373                         raise_from(ExceptionType(msg), None)
    374 
--> 375                 return func(self, **parsed['args'])
    376         else:
    377             def func_call(*args, **kwargs):

~/Src/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
   1377                              % str(container.__class__.__name__))  # noqa: F821
   1378         else:
-> 1379             return attr_map.construct(builder, build_manager)
   1380 
   1381     @docval({"name": "container", "type": Container, "doc": "the container to convert to a Builder"},

~/Src/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    373                         raise_from(ExceptionType(msg), None)
    374 
--> 375                 return func(self, **parsed['args'])
    376         else:
    377             def func_call(*args, **kwargs):

~/Src/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    960         except Exception as ex:
    961             msg = 'Could not construct %s object' % (cls.__name__,)
--> 962             raise_from(Exception(msg), ex)
    963         return obj
    964 

~/conda/envs/pynwb/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

Exception: Could not construct ElectricalSeries object
```
</p></details>

### Environment
Please describe your environment according to the following bullet points.
- **Python Executable:** Conda 
- **Python Version:**  Python 3.6
- **Operating System:**  macOS 
- **Pynwb Version:** Latest dev (https://github.com/NeurodataWithoutBorders/pynwb/commit/426109b85c2e683d96f2126286fb3369013abb8b)

## Checklist

- [x] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?

**Comment by tjd2002** [comment]
Just realized this is a regression. Running the sample script using `git bisect`* identifies c55ff32e2abe1b0481668f5217d21f822639ea43 as the place where this bug was introduced. CC @ajtritt 

```
c55ff32e2abe1b0481668f5217d21f822639ea43 is the first bad commit
commit c55ff32e2abe1b0481668f5217d21f822639ea43
Author: Andrew Tritt <ajtritt@lbl.gov>
Date:   Mon Oct 8 21:04:55 2018 -0700

    refactor __get_subspec_values

:040000 040000 aa4145284a4f8ff2b51e2b149e4026ee1a906e1e cb0ac855f01c186e434a62021131678ffba90fc6 M	src
```

*Man do I love git bisect!
```
git bisect start dev 430ae547
git bisect run ETR_test_679_fails.py # test script from above
```

**Comment by ajtritt** [comment]
@tjd2002 PR #682 includes a check for this. This fix is a bit of a work around rather than a general solution. A general solution to this should be addressed by issue #633
